### ISSUE TITLE
FE #508-defect-2 [COGNITION, SEMANTIC MARKUP]: Form field instructions should be descriptive and MUST be associated with inputs and buttons

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#19961]
# [508-defect-2](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-2)

## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

## Point of Contact

**VFS Point of Contact:** _Josh_

## User Story or Problem Statement

As a screen reader user, I want to know which fields are grouped together so that I can have an equivalent experience to sighted users when navigating my direct deposit options.

## Details

> Using the right HTML elements when implementing forms is essential to ensure they can be used by as many people as possible including screen reader users.

Sighted users will be able to _visually_ determine that fields are grouped together thanks to the grey border. However, screen reader users will not have access to the same information for the following reasons:
- Upon clicking "edit," screen reader users may likely tab to the first item expecting a form field. Screen reader users will hear "Where can I find these numbers," which has no meaning on its own without a group label (What numbers?). The first set of instructions will be skipped, and will not be described to the user upon landing on the help accordion.
- Upon tabbing again and entering the first input field, users will hear "Routing number (Your bank’s name will appear after you add the 9-digit routing number)," but no other information to help contextualize the task. Screen reader users with cognitive disabilities may struggle to remember why they are providing this information.
- Upon tabbing again, users will hear "Account number (This should be no more than 17 digits)(*Required)." Screen reader users may struggle to understand that this is grouped with the first input field as they would not have heard the initial instructions "Please enter your bank’s routing and account numbers and your account type" while tabbing.

## Acceptance Criteria

- [ ] Fields are semantically grouped and labelled
- [ ] Instructions are descriptive

## Environment

* Operating System: any
* Browser: all
* Screenreading device: any
* Server destination: all

## Steps to Recreate

1. Enter `https://staging.va.gov/profile` in browser
2. Start any screenreader 
3. Tab to the first edit button and select it
4. Confirm the screen reader announces some variation of "Where can I find these numbers, button"-- but _does not_ announce a group label (e.g. edit bank account for disability compensation and pension benefits)

## Ideal Solution

Ideally, a preferred semantic approach to solving this problem would be to, upon activating edit mode,
- extend the `form` to before the `dfn` "Account" in the DOM
- wrap the `form` contents within a `fieldset`
- convert the `dfn` "Account" to a `legend` labelled "Edit bank account for disability compensation and pension benefits"
- ensure the `legend` is the first child of the `fieldset`

```html
<form>
  <fieldset>
    <legend> <!-- Must be the first child in order to be announced -->
      Edit bank account
    </legend
    <!-- fields go here -->
  </fieldset>
</form>
```

This would add a semantic group label to all fields within the fieldset. If a screen reader were to tab to the first button "Where can I find these numbers..." the screen reader would also announce, "edit bank account" which contextualizes its use (bonus points if we can rewrite "where can I find these numbers" to "where can I find my bank account numbers" to meet link purpose). This would apply to all other fields, including the update and cancel buttons. We would not need to rely on `aria`, which meets the first rule of `aria` which is to _not_ use `aria` when it can be avoided with pure semantic html.

## Band Aid Solution

That being said, the ideal solution may require extensive changes to the form design system. An alternative, and less change-heavy method of meeting the acceptance criteria _might_ be to use `role="group"` and `aria-label`, but more testing would be needed.

```html
<div role="group" aria-label="Edit bank account for disability compensation and pension benefits">
  <span>
    <div id="errors"></div>
    <p class="vads-u-margin-top--0">Please enter your bank’s routing and account numbers and your account type.</p>
    <!-- etc -->
  </span>
</div>
```

## WCAG or Vendor Guidance (optional)

* [Using the fieldset and legend elements, gov.uk](https://accessibility.blog.gov.uk/2016/07/22/using-the-fieldset-and-legend-elements/)
* [Grouping controls, WCAG guidance](https://www.w3.org/WAI/tutorials/forms/grouping/)

### Existing Issues
<img width="1792" alt="Screen Shot 2021-02-05 at 1 51 08 PM" src="https://user-images.githubusercontent.com/14154792/107076821-f8759c00-67b9-11eb-9a66-300a887b53cc.png">
<img width="1792" alt="Screen Shot 2021-02-05 at 1 41 31 PM" src="https://user-images.githubusercontent.com/14154792/107076824-fad7f600-67b9-11eb-817d-9be18f921b65.png">

### Proposed Ideal Solution 
<img width="1717" alt="Screen Shot 2021-02-05 at 2 37 09 PM" src="https://user-images.githubusercontent.com/14154792/107080852-d0893700-67bf-11eb-86a0-3147153221d4.png">

https://user-images.githubusercontent.com/14154792/107080854-d252fa80-67bf-11eb-831a-083a019d6dc2.mov

**Comment by mshea0606** [comment]
@1Copenut is this a form systems issue? cc @erikphansen @mshea0606 @sporeboy

**Comment by joshkimux** [comment]
@erikphansen Let me know when y'all start up work on this again and/or have any questions on the ticket. I'm not certain if this is a forms system issue, but if it is, I'm more than happy to help explore alternatives to get this closed out 😄

**Comment by joshkimux** [comment]
@artsymartha68 flagging this up to you in case you might know who currently owns (if anyone) direct deposit.

**Comment by adamwhitlock1** [comment]
I'm having trouble coming up with a solution that works for this issue. Since the form system is rendering this set of fields, there isn't a great way to do this in the form system itself.

When I tried the bandaid solution using `<div role="group" aria-label="Edit bank account for disability compensation and pension benefits">` to wrap the form, the group worked in describing the additional info component in Voiceover but the subsequent input fields did not have the group label read and would just say routing number or account number.

**Comment by andaleliz** [comment]
@adamwhitlock1 that's awesome that we added the aria labels for the forms! 🎉 that'll be a big improvement.

I'm curious what you and @joshkimux think of a re-organization of elements as a possible solution. Instead of the current order of elements on the page, what if we ordered the fields as:
- account type
- routing number
- account number

and then put the FAQ below the fields, and above the button?

I honestly think it makes a little more sense to say what kind of account it is first anyway.  In other examples I've seen on VA.gov, the additional info component comes _after_ the related questions, so I think this would also be more aligned w/ how it's used elsewhere on the site as well.  

Here's a visual of what I'm proposing:

<img width="330" alt="image" src="https://github.com/department-of-veterans-affairs/va.gov-team/assets/55992116/89740370-d564-4a3d-9aba-c37b2f5622db">

I'm not sure if that totally solves the issue with the form system though. :/

**Comment by adamwhitlock1** [comment]
@andaleliz that is an interesting idea, and I'm down to try anything! Yea it doesn't quite fix the accessibility issue, but it kind of seems like a clearer ordering to me.

**Comment by andaleliz** [comment]
Oh but wait! It sounds like with the bandaid solution, the additional info component is interrupting the flow - so maybe moving that component below the fields _would_ fix it, because it would read the label and then the subsequent fields? Or am I missing some other part of it?

>When I tried the bandaid solution using `<div role="group" aria-label="Edit bank account for disability compensation and pension benefits">` to wrap the form, the group worked in describing the additional info component in Voiceover but the subsequent input fields did not have the group label read and would just say routing number or account number.

**Comment by adamwhitlock1** [comment]
That is a good point! @mtcA6 maybe we roll this one over for a point or so, and I could try doing that reordering (probably quick and dirty in the code at first locally) then see if that helps, and we could maybe go that route?

**Comment by andaleliz** [comment]
@adamwhitlock1 quick n dirty sounds perfect - there is unclear guidance on this in the design system that I'm trying to get clarity on ([my question in Slack](https://dsva.slack.com/archives/C01DBGX4P45/p1693341374910579)). If they have no answer, then I will need to get it reviewed in content office hours, which I could do this Thursday.  I'll report back, and also stay tuned for the outcome of your testing.

**Comment by adamwhitlock1** [comment]
@mtcA6 I would just carry over 1 or 2 points total. I did the majority of the work to see what can be done, and am now going to try a 3rd option locally to see if it could be a potential solution.

**Comment by andaleliz** [comment]
Reporting back:

- The design system team doesn't have official guidance on the order of fields for this pattern
- The content folks agree with the proposed solution to reorder the fields and move the FAQ to the bottom of the form.

@adamwhitlock1 if you feel this solves the a11y issue, we should move forward. Let me know.

**Comment by adamwhitlock1** [comment]
Okay, so travis and I talked, and I think we are going to try and get the fields reversed and see if that is acceptable as far as screen readers and making it easier to understand what direct deposit section is being updated. I will make a ticket specifically for that work and link to this ticket and the convo on this.

**Comment by andaleliz** [comment]
@adamwhitlock1 are you going to take it to a11y office hours, or have some other kind of review to see if that's acceptable? If so, can I join please? I'd love to learn.

**Comment by adamwhitlock1** [comment]
@andaleliz yep! I'll make sure you are invited to the meeting/office hours when I'm ready to have it reviewed. https://github.com/department-of-veterans-affairs/va.gov-team/issues/65506 <- here is the ticket for the work itself

**Comment by mtcA6** [comment]
Closing this as we're using the [workaround solution here](https://github.com/department-of-veterans-affairs/va.gov-team/issues/65506)
